modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -146,18 +172,53 @@ public void deactivate() {
 
             updaterTread = null;
         }
+
+        leaseTracker.unsubscribeLeaseAdded();
+
+        leaseConciliator = null;
     }
 
     /**
      * Finds a node that can be the leaseholder.
      *
+     * @param groupId Replication group id.
      * @param assignments Replication group assignment.
      * @return Cluster node, or {@code null} if no node in assignments can be the leaseholder.
      */
-    private ClusterNode nextLeaseHolder(Set<Assignment> assignments) {
+    private ClusterNode nextLeaseHolder(ReplicationGroupId groupId, Set<Assignment> assignments) {
         //TODO: IGNITE-18879 Implement more intellectual algorithm to choose a node.
+        LeaseAgreement agreement = leaseConciliator.conciliated(groupId);
+
+        if (agreement.getRedirectTo() != null) {
+            boolean hasInAssignments = false;
+
+            for (Assignment assignment : assignments) {
+                if (agreement.getRedirectTo().equals(assignment.consistentId())) {
+                    hasInAssignments = true;
+
+                    break;
+                }
+            }
+
+            if (hasInAssignments) {
+                ClusterNode candidate = topologyTracker.nodeByConsistentId(agreement.getRedirectTo());
+
+                if (candidate != null) {
+                    return candidate;
+                }
+            }
+        }
```

Comment:
I strongly doubt that this should be handled here. I would mention the interface that I have suggested in https://issues.apache.org/jira/browse/IGNITE-18879 : on this stage we need only candidate for a group, without consideration of redirect proposal. Moreover, I see no reason to store intermediate state with redirect, we can only remember that we have chosen the candidate (and that candidate was saved to meta storage) and redirect proposal can be handled reactively.

Additionally (and it is not only about this part of code) the implementation seems too complicated for me and definitely lacks reactiveness. We have updater scheduler, map with leases, map with lease agreements, metastorage update listeners that can change these maps, several intermediate states, etc. - we will have pretty long explanation just to describe all of the logic that is contained here. I suggest to think about how to simplify it.

Reply:
Without the logic, I cannot write a test. If it is now, it does not mean that it cannot be rewritten in the future.